README.md
# i18n-pseudo

[Pseudotranslation](https://en.wikipedia.org/wiki/Pseudotranslation) is an incredibly useful tool for localizing your apps. This module makes it easy to apply pseudo to any given string.

If the provided string is an ICU Message formatted string, it will parse and generate the ICU Message with proper pseudo.

# Installation

```shell
$ npm i i18n-pseudo
```

or for you `yarn` fans

```shell
$ yarn add i18n-pseudo
```

or using `npx`:
```
$ npx i18n-pseudo "Hello, World"
你好Ĥệḻḻỗ, 𝕎ỗřḻḋ àḃċḋệḟĝȟíǰ世界
```

# CLI

## Example

```shell
$ pseudo Awesome Text Here
你好Ã𝚠ệṧỗṃệ Ťệẍẗ Ĥệřệ àḃċḋệḟĝȟíǰǩḻṃņ世界
```

And if you don't want to expand the tet

```shell
$ pseudo Awesome Text Here -ne
世界Ã𝚠ệṧỗṃệ Ťệẍẗ Ĥệřệ你好
```

## Options

| Argument            | Default | Description |
| ------------------- | ------- | ----------- |
| `--noExpand`, `-ne` | `false` | When provided, the script will _not_ expand the incoming text |

# Code

## Example

Basic usage

```javascript
import { PseudoFormat } from 'i18n-pseudo';
const genPseudo = new PseudoFormat();
console.log(genPseudo.format("Awesome Text Here"));
// 你好Ã𝚠ệṧỗṃệ Ťệẍẗ Ĥệřệ àḃċḋệḟĝȟíǰǩḻṃņ世界
```

It can apply pseudo to ICU Messages as well:

```javascript
import { PseudoFormat, PseudoFormatOptions } from 'i18n-pseudo';

const text = `You have {count, plural,
    zero { nothing }
    one { # item }
    few { # items }
    many { # items }
} in your cart`;

const opts:Partial<PseudoFormatOptions> = { doExpand: false };
const genPseudo = new PseudoFormat(opts);
console.log(genPseudo.format(text));

/*
世界Ỹỗű ȟàṿệ { count, plural,
  zero { ņỗẗȟíņĝ }
  one { # íẗệṃ }
  few { # íẗệṃṧ }
  many { # íẗệṃṧ }
} íņ ẙỗűř ċàřẗȟṔṃỗŚqņṼřṿǰẗÃÃȐṿṃÃȐḋŚǰqḃ你好
*/
```

## Options

### `PseudoFormatOptions`

| Property      | Default | Description |
| ------------- | ------- | ----------- |
| `appendChars` | `"你好"` | When provided, these characters will be appended to the output. This is helpful when you have certain characters that always seem to give your system trouble
| `doExpand`    | `true`  | When `false`, will _not_ expand the input. Expanding characters is useful to ensure your codebase accounts for move verbose languages
| `expandChars`    | ASCII upper and lower chars  | A string of characters that will be randomly selected to create an expansion of the input string.
| `prependChars` | `"世界"` | When provided, these characters will be prepended to the output. This is helpful when you have certain characters that always seem to give your system trouble
| `pseudoChars` | [Reference](https://github.com/AaronPresley/i18n-pseudo-js/blob/master/src/char-map.ts) | A hash map of the pseudo characters that will replace the ASCII characters. For default value, see the `pseudoCharacterMap` value in the `char-map.ts` file

CHANGELOG.md
# Changelog

## `v1.1.3`

My initial text expansion seemed a bit excessive, so I reduced to the following:

| Char Length | Previous Expansion | Updated Expansion |
| ----------- | ------------------ | ----------------- |
| 0 thru 5    | 120%               | 100%              |
| 6 thru 10   | 100%               | 60%               |
| 11 thru 25  | 80%                | 40%               |
| 26 thru 50  | 60%                | 20%               |
| 51 thru 75  | 40%                | 0%                |
| 51+         | 20%                | 0%                |

Upcoming improvements will allow setting this expansion percentage through options. This will allow users to overwrite these defaults.

src/pseudo-translate.ts
import { ExpandTextOptions, PseudoTranslateOptions } from "./types";
import { pseudoCharacterMap, ignoreCharacterMap, expansionCharacters } from './char-map';

const DEFAULT_EXPAND_OPTIONS:ExpandTextOptions = {
  expansionPercentage: 'auto',
  possibleCharacters: expansionCharacters,
};

const DEFAULT_PSEUDO_OPTIONS:PseudoTranslateOptions = {
  appendCharacters: '你好',
  expandText: true,
  prependCharacters: '世界',
};

/**
 * A function that expands the length of a given string
 * @param input The string to expand
 * @param options The options, has reasonable defaults
 */
export const expandText = (
  input:string,
  options: ExpandTextOptions = DEFAULT_EXPAND_OPTIONS,
):string => {
  const finalOpts = { ...DEFAULT_EXPAND_OPTIONS, ...options };
  const { expansionPercentage, possibleCharacters } = finalOpts;
  
  let output = `${input}`;
  let finalExpansionPercent = expansionPercentage;

  // Determine our expansion amount based on the length of the input string
  // but only if we're set to auto expand
  if (finalExpansionPercent === 'auto') {
    if (input.length <= 5) finalExpansionPercent = 120
    else if (input.length <= 10) finalExpansionPercent = 100
    else if (input.length <= 25) finalExpansionPercent = 80
    else if (input.length <= 50) finalExpansionPercent = 60
    else if (input.length <= 75) finalExpansionPercent = 40
    else finalExpansionPercent = 20
  }

  // Figure out how many characters
  // const addLength = Math.round(input.length * (finalExpansionPercent / 100)) - possibleCharacters.length;
  const addLength = Math.round(input.length * (finalExpansionPercent / 100));

  // Throw an error if we're supposed to add more characters than we were provided
  // Maybe eventually we should just choose random characters within possibleCharacters?
  if (addLength && addLength > possibleCharacters.length)
    throw new Error(`I'm supposed to expand this string by ${addLength} characters but the possibleCharacters option only has ${possibleCharacters.length} characters available`);
  
  // Append the characters and return
  return `${output} ${possibleCharacters.substr(0, addLength)}`;
};


/**
 * A function that translates a normal ASCII string to a pseduotranslated string
 * @param input The ASCII string to pseudotranslate
 * @param options The options, has reasonable defaults
 */
export const pseudoTranslate = (
  input:string,
  options: PseudoTranslateOptions = DEFAULT_PSEUDO_OPTIONS,
):string => {
  const finalOpts = { ...DEFAULT_PSEUDO_OPTIONS, ...options };
  const { appendCharacters, prependCharacters, expandText: doExpandText } = finalOpts;

  let finalInput = input;

  if (doExpandText) finalInput = expandText(finalInput);

  let ignoreUntilChar:string = null;
  let output:string = '';

  // Loop through each character of this string
  for (let x = 0; x < finalInput.length; x += 1) {
    const char = finalInput[x];

    // Does this character match one of our ignore characters?
    if (Object.keys(ignoreCharacterMap).includes(char))
      ignoreUntilChar = ignoreCharacterMap[char];

    // Stop ignoring characters
    if (!!ignoreUntilChar && char === ignoreUntilChar)
      ignoreUntilChar = null;

    
    output += !!ignoreUntilChar
      ? char // We're ignoring chars, so append as-is
      : (pseudoCharacterMap[char] || char); // Append the pseudo char with a fallback
  }
  
  return `${appendCharacters}${output}${prependCharacters}`;
};
